lib/puppet/util/execution.rb

```diff
@@ -89,6 +89,47 @@ def self.execpipe(command, failonfail = true)
     output
   end
 
+  # Wait for a process to exit or kill it after a given timeout
+  #
+  # This function polls for the exit status of a child process. If the
+  # supplied timeout is exceeded, then the child is terminated.
+  #
+  # @param child_pid [Integer] The PID of the process to monitor.
+  # @param timeout [Integer] The maximum amount of time to wait, in seconds.
+  #
+  # @return [Array<Integer, Process::Status>] A two element array containing
+  #   the PID of the monitored process along with a Process::Status object
+  #   containing its exit status as returned by Process.waitpid2.
+  def self.wait_with_timeout(child_pid, timeout)
+    # TODO: Use a monotonic clock.
+    deadline = Time.now + timeout
+    # The Process module on Windows only supports :KILL, unless a win32
+    # extension gem is used. Even then, the :TERM supplied is equivalent
+    # to :KILL
+    kill_signal = Puppet.features.microsoft_windows? ? :KILL : :TERM
+
+    loop do
+      # Poll status of child process without blocking. Returns nil
+      # if the process has not exited.
+      status = Process.waitpid2(child_pid, Process::WNOHANG)
+
+      return status unless status.nil?
+
+      # TODO: Use a monotonic clock.
+      if Time.now > deadline
+        Puppet.err(_('Child process exceded timeout. Sending %{signal} signal to PID %{pid}') %
+                   {pid: child_pid,
+                    signal: kill_signal})
+
+        Process.kill(kill_signal, child_pid)
```

Comment:
Yeah. There's a comment in there about `:TERM` being supported if win32-process is present. I stuck with switching to `:KILL` because there doesn't seem to be a difference between how win32-process implements the two signals, so it seemed better to avoid taking on a dependency if no benefit is provided.